### ISSUE TITLE
process_teardown() hang

### ISSUE THREAD

**Issue by jonthn** [neovim/neovim#6891]
- `nvim --version`:

NVIM v0.2.1-331-g6650588c4
Build type: Debug
Compilation: clang -O2 -I/usr/include -I~/.pkg/dist/include -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS
Compiled by jonthn@mac.local

Optional features included (+) or not (-): +acl   +iconv    -jemalloc +tui
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "~/.pkg/dist/share/nvim"

(slightly edited to remove some of my local path)

- Operating system/version:

macOS 10.12.4

- Terminal name/version:

iTerm 2 - 3.0.15

- `$TERM`:

xterm-256color

### Actual behaviour

nvim starts without problems but after entering `:q` it exists and right before giving me the access to my shell it stays there forever

### Expected behaviour

It should quit normally

### Steps to reproduce using `nvim -u NORC`

It can't be reproduced with NORC see below, it's reproducible (at least in my case) when running/using plugin `vim-signify`

### Explanations / troubleshooting done

I attached (using lldb) to the process while it was hanging and it was in : 

process.c > process_teardown(Loop *loop) > LOOP_PROCESS_EVENTS_UNTIL

nvim is waiting for a process to finish unfortunately this process doesn't exist anymore, so I tried (not in a proper way if you look at the diff) to fix it by catching the `UV_ESRCH` return code from uv_kill() and remove this child from the loop but fails later() usually with a segfault.

So I tried to look how nvim can be in such state and I believe it's because the process is started before the main loop, and because of the quick exit from the sub-process isn't detected and not handled as it should.
In the diff : https://gist.github.com/jonthn/5e34b7b516b6297386464de9ffcddf24

You'll see the debug code I added and below you can take a look at the output log : 

https://gist.github.com/jonthn/1fd14f99fa5a670d29f586c9f0f49c07

**Comment by justinmk** [comment]
@jonthn Thanks for the great debugging! In `process_teardown`, what happens if you change the `-1` arg to `LOOP_PROCESS_EVENTS_UNTIL` to 2000 (=2 seconds)? Any segfaults? Non-zero exit code? "Vim caught deadly signal message"? 

In https://github.com/neovim/neovim/pull/6827 I added a timeout (and logging) to `loop_close`, maybe we need that in `loop_poll_events()` as well.

> process.c > process_teardown(Loop *loop) > LOOP_PROCESS_EVENTS_UNTIL

Is it hanging in `loop_poll_events`? What is the full stacktrace?  (Line numbers would also help if possible.)

> I believe it's because the process is started before the main loop, and because of the quick exit from the sub-process isn't detected and not handled as it should.

main loop shouldn't matter, but "quick exit" process is similar to what I and @blueyed concluded when investigating https://github.com/neovim/neovim/pull/6823

**Comment by blueyed** [comment]
@jonthn 
Would be great if you could test this with https://github.com/neovim/neovim/pull/6823 indeed.

**Comment by jamessan** [comment]
I wonder if that's also related to what I saw in my [PR](https://github.com/neovim/neovim/pull/6195#issuecomment-289188508).  It's also quickly spawning/killing processes.

**Comment by jonthn** [comment]
@justinmk 

* I forgot to tell you that when I start `vim`I have this message : 
```
Vim(cscope):cs_read_prompt EOF: Interrupted system call
```
* Using 2 seconds timeout doesn't solve the problem
* I can see the message below when I try to kill the `vim` process from another shell

`Vim: Caught deadly signal 'SIGTERM'
Vim: Finished.
nvim(68676,0x7fff9f6da3c0) malloc: *** error for object 0x7fa2075ef440: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug` 

* The full stacktrace when it's hanging (without timeout)

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff9689ad96 libsystem_kernel.dylib`kevent + 10
    frame #1: 0x000000010dea80cf libuv.1.dylib`uv__io_poll + 735
    frame #2: 0x000000010de97598 libuv.1.dylib`uv_run + 456
    frame #3: 0x000000010dbd9a47 nvim`loop_poll_events(loop=0x000000010ddcf988, ms=<unavailable>) at loop.c:52 [opt]
    frame #4: 0x000000010dbdab2d nvim`process_teardown(loop=0x000000010ddcf988) at process.c:135 [opt]
    frame #5: 0x000000010dc50aa9 nvim`event_teardown at main.c:166 [opt]
    frame #6: 0x000000010dcb0069 nvim`mch_exit(r=1) at os_unix.c:144 [opt]
    frame #7: 0x000000010dc534a9 nvim`getout(exitval=<unavailable>) at main.c:644 [opt]
    frame #8: 0x000000010dc00e5a nvim`ex_quit(eap=0x00007fff5209ec60) at ex_docmd.c:5980 [opt]
    frame #9: 0x000000010dbf4fb6 nvim`do_one_cmd(cmdlinep=0x00007fff5209ef60, flags=0, cstack=<unavailable>, fgetline=<unavailable>, cookie=0x0000000000000000) at ex_docmd.c:2241 [opt]
    frame #10: 0x000000010dbf0dec nvim`do_cmdline(cmdline=<unavailable>, fgetline=<unavailable>, cookie=<unavailable>, flags=0) at ex_docmd.c:607 [opt]
    frame #11: 0x000000010dc88177 nvim`nv_colon(cap=0x00007fff5209f5f8) at normal.c:4499 [opt]
    frame #12: 0x000000010dc85db5 nvim`normal_execute(state=0x00007fff5209f558, key=<unavailable>) at normal.c:1137 [opt]
    frame #13: 0x000000010dd3357f nvim`state_enter(s=0x00007fff5209f558) at state.c:61 [opt]
    frame #14: 0x000000010dc7fcb5 nvim`normal_enter(cmdwin=<unavailable>, noexmode=<unavailable>) at normal.c:467 [opt]
    frame #15: 0x000000010dc52a20 nvim`main(argc=<unavailable>, argv=<unavailable>) at main.c:553 [opt]
    frame #16: 0x00007fff9676b235 libdyld.dylib`start + 1
    frame #17: 0x00007fff9676b235 libdyld.dylib`start + 1
```

* The full stacktrace when it's hanging (with the 2 seconds timeout)

```
* frame #0: 0x00007fff9689ad96 libsystem_kernel.dylib`kevent + 10
    frame #1: 0x00000001071820cf libuv.1.dylib`uv__io_poll + 735
    frame #2: 0x0000000107171598 libuv.1.dylib`uv_run + 456
    frame #3: 0x0000000106ebab6a nvim`loop_close(loop=0x00000001070b0988, wait=<unavailable>) at loop.c:93 [opt]
    frame #4: 0x0000000106f31aca nvim`event_teardown at main.c:172 [opt]
    frame #5: 0x0000000106f91069 nvim`mch_exit(r=1) at os_unix.c:144 [opt]
    frame #6: 0x0000000106f344a9 nvim`getout(exitval=<unavailable>) at main.c:644 [opt]
    frame #7: 0x0000000106ee1e5a nvim`ex_quit(eap=0x00007fff58dbdc60) at ex_docmd.c:5980 [opt]
    frame #8: 0x0000000106ed5fb6 nvim`do_one_cmd(cmdlinep=0x00007fff58dbdf60, flags=0, cstack=<unavailable>, fgetline=<unavailable>, cookie=0x0000000000000000) at ex_docmd.c:2241 [opt]
    frame #9: 0x0000000106ed1dec nvim`do_cmdline(cmdline=<unavailable>, fgetline=<unavailable>, cookie=<unavailable>, flags=0) at ex_docmd.c:607 [opt]
    frame #10: 0x0000000106f69177 nvim`nv_colon(cap=0x00007fff58dbe5f8) at normal.c:4499 [opt]
    frame #11: 0x0000000106f66db5 nvim`normal_execute(state=0x00007fff58dbe558, key=<unavailable>) at normal.c:1137 [opt]
    frame #12: 0x000000010701457f nvim`state_enter(s=0x00007fff58dbe558) at state.c:61 [opt]
    frame #13: 0x0000000106f60cb5 nvim`normal_enter(cmdwin=<unavailable>, noexmode=<unavailable>) at normal.c:467 [opt]
    frame #14: 0x0000000106f33a20 nvim`main(argc=<unavailable>, argv=<unavailable>) at main.c:553 [opt]
    frame #15: 0x00007fff9676b235 libdyld.dylib`start + 1
    frame #16: 0x00007fff9676b235 libdyld.dylib`start + 1
```

@blueyed 

With the patch from #6823, still the same behaviour.
Stacktrace : 
```
  * frame #0: 0x00007fff9689ad96 libsystem_kernel.dylib`kevent + 10
    frame #1: 0x000000010fb980cf libuv.1.dylib`uv__io_poll + 735
    frame #2: 0x000000010fb87598 libuv.1.dylib`uv_run + 456
    frame #3: 0x000000010f8cd9d7 nvim`loop_poll_events(loop=0x000000010fac3988, ms=<unavailable>) at loop.c:52 [opt]
    frame #4: 0x000000010f8cea9d nvim`process_teardown(loop=0x000000010fac3988) at process.c:133 [opt]
    frame #5: 0x000000010f944a89 nvim`event_teardown at main.c:166 [opt]
    frame #6: 0x000000010f9a4049 nvim`mch_exit(r=1) at os_unix.c:144 [opt]
    frame #7: 0x000000010f947489 nvim`getout(exitval=<unavailable>) at main.c:644 [opt]
    frame #8: 0x000000010f8f4e3a nvim`ex_quit(eap=0x00007fff503aac60) at ex_docmd.c:5980 [opt]
    frame #9: 0x000000010f8e8f96 nvim`do_one_cmd(cmdlinep=0x00007fff503aaf60, flags=0, cstack=<unavailable>, fgetline=<unavailable>, cookie=0x0000000000000000) at ex_docmd.c:2241 [opt]
    frame #10: 0x000000010f8e4dcc nvim`do_cmdline(cmdline=<unavailable>, fgetline=<unavailable>, cookie=<unavailable>, flags=0) at ex_docmd.c:607 [opt]
    frame #11: 0x000000010f97c157 nvim`nv_colon(cap=0x00007fff503ab5f8) at normal.c:4499 [opt]
    frame #12: 0x000000010f979d95 nvim`normal_execute(state=0x00007fff503ab558, key=<unavailable>) at normal.c:1137 [opt]
    frame #13: 0x000000010fa2755f nvim`state_enter(s=0x00007fff503ab558) at state.c:61 [opt]
    frame #14: 0x000000010f973c95 nvim`normal_enter(cmdwin=<unavailable>, noexmode=<unavailable>) at normal.c:467 [opt]
    frame #15: 0x000000010f946a00 nvim`main(argc=<unavailable>, argv=<unavailable>) at main.c:553 [opt]
    frame #16: 0x00007fff9676b235 libdyld.dylib`start + 1
    frame #17: 0x00007fff9676b235 libdyld.dylib`start + 1
```

but the NVIM_LOG_FILE gives us : 

```
2017/06/14 15:54:06 INFO  69271/main:548: starting main loop
2017/06/14 15:54:08 INFO  69271/process_stop:239: Sending SIGTERM to pid 69273
2017/06/14 15:54:08 INFO  69271/process_stop:255: Starting job kill timer
2017/06/14 15:54:10 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:12 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:14 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:16 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:18 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:20 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:22 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:24 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:26 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:28 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:30 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:32 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:34 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:36 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:38 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:40 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:42 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:44 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
2017/06/14 15:54:46 INFO  69271/children_kill_cb:280: Sending SIGKILL to pid 69273
```

I hope it gives your more informations.
If you need anything else or if you want to try something else don't hesitate.

**Comment by justinmk** [comment]
This is good info. During the hang, Can you check if the process we are attempting to SIGKILL exists?

**Comment by jonthn** [comment]
The process isn't running anymore. It's already terminated

**Comment by justinmk** [comment]
> Vim(cscope):cs_read_prompt EOF: Interrupted system call

Should be fixed by #8079 (and #7633).
